### DISCUSSION TITLE
mix logger.error(String format, Object... arguments) and logger.error(String msg, Throwable t)

### DISCUSSION THREAD

**Discussion by ManfredSchenkIOSB** [qos-ch/slf4j#296] [Ideas]
from time to time I am missing some kind of combination of the methods mentioned in the discussion title. Within a catch clause I'm trying to log some parameters as well as the Exception. At the moment, I construct the msg text by myself, but a mix of the methods would be nice. The same is true for all methods taking a Throwable as parameter.

But I see there are points in favor of such a method and also against this method:

positive:
- similarity (You can always use the formatter approach)
- reduced overhead when Level is not enabled

negative:
- Breaking API change ( the Throwable would have to be the first parameter to be compatible with the Object... parameter)
- no real performance gain for the use case error (The Level error will probably be active in nearly all situations, and Throwables are most likely logged with Level Error)

Any opinions on this?

**Comment by ceki** [comment]
@ManfredSchenkIOSB  Are you aware of [this faq entry](https://www.slf4j.org/faq.html#paramException)?

Note that starting with 2.0.0, you can also write
```
 try {
 } catch(Exception ex) {
   logger.atError().setMessage("Hello {}").addArgument("world").setCause(ex).log();
 }

**Reply by ManfredSchenkIOSB** [reply]
Now I am aware of the faq entry :-) . But if I understand correctly, the faq entry only handles the case where there is only one additional parameter besides the Throwable.

**Reply by ceki** [reply]
Actually, any number of parameters, including 0, 1 or more.

**Comment by JacksonBaileyHM** [comment]
Unless I am misunderstanding you, this is actually already possible. [See here.](https://www.slf4j.org/faq.html#paramException
) Basically when the last argument is a `Throwable` and doesn't match to any `{}` in the format string then it treats it similarly to `error(String, Throwable)`.

```java
String s = "Hello world";
try {
  Integer i = Integer.valueOf(s);
} catch (NumberFormatException e) {
  logger.error("Failed to format {}", s, e);
}
```

It would be nice if this were clearly mentioned in the Javadoc on the methods though. I forget it often myself. (In fact that's why I stumbled here, wondering if it could be added.) It is (in my opinion, cryptically) mentioned [here](https://github.com/qos-ch/slf4j/blob/7c1d86e21e7260093bbef5647cd554e09d64f68b/slf4j-api/src/main/java/org/slf4j/Logger.java#L76) in the Javadoc.

**Comment by q3769** [comment]
> The SLF4J API supports parametrization (of 1 or more objects) in the presence of an exception, assuming the exception is the last parameter.

Sorry but that is a bad idea for a strongly-typed language like Java. Giving special treatment for an argument simply because of its position in a method's signature is a convention-based approach that should have been replaced by one that is based on the argument's type.

Why not simply put the "Throwable" type as the very first argument in the Logger.log(Throwable t, String message, Object... args) signature? It seems to be straight-forward API design 101. What am I missing?

Yes, the fluent API in version 2.x is nice but that doesn't solve the earlier bad API design as-is.

**Reply by JacksonBaileyHM** [reply]
>Why not simply put the "Throwable" type as the very first argument in the Logger.log(Throwable t, String message, Object... args) signature? 

Yes, that would be fine.

> What am I missing? Yes, the fluent API in version 2.x is nice but that doesn't solve the earlier bad API design as-is.

That changing the behavior of the `log(String, Object...)` methods would be a backwards incompatible change. Avoiding those is also API design 101, no? 😉 I see your point though, it is sort of odd, but it is what it is. I just wish it were documented a little better. I've had multiple teammates use things like `error("Problem with input '{}', exception: {}", input, e.getMessage())` as opposed to `error("Problem with input '{}'", input, e)`. Based on my experience (with an incredibly small sample size) I think folks just don't know about it.

**Reply by q3769** [reply]
Throwable being the first arg is a new API signature, there is no backward compatible issue. 

Semantically you could still keep the log(String, Object...) behavior although I wouldn't - for the reasons mentioned above. Actually, specially treating the last arg if it is a Throwable was a semantic "breaking change" in the first place. 

The whole backward compatible thing is nice but is really going too far with Java - what is the point of being backward compatible with a bad idea (the jdk Date/Calendar API prior to Java 8, the JDK1.4 logging, ...)? In this case the API is versioned anyway, people can stay at a lower version forever if they so choose.

**Reply by prdoyle** [reply]
I'm a fan of strong types, but the badness of the current API is being overstated.

A few facts I think we can agree on, for context:
- Java has positional parameters; it's the nature of the language.
- If the types overlap, you can accidentally swap two of the parameters and encounter a problem. This is also the nature of the language. 
- Log formatting is sensitive to parameter order regardless of exceptions. Swap the order and your log message will look funny.

What you are proposing might help the compiler report a situation where someone meant to have an exception provided as the _cause_ but instead printed it into their format string. It would catch _only_ this case, and no other case of log parameters having the wrong order. And when this case occurs, the application will not experience any errors or other bad behaviour; the log entries will just be missing some information.

Yes, it's conceivable that your change might theoretically catch the occasional bug whose consequences are relatively minor. The cost of your change is that we will have, for the first time ever, a logging method with positional arguments whose first argument is not the message string. This will look odd and counterintuitive to a lot of users. Yes, this cost is low, but so is the benefit, so it can't be ignored.

The fluent API already fixes this anyway.

**Reply by q3769** [reply]
Most oo languages have the positional param feature, but that doesn't mean it's the good way to do everything. On the contrary, using the strong type feature over position is a, if not the, big point of oo. That's the reason for the proposed signature.

The argument against "for the first time" is really hard to understand. The way things get done should depend on the merit of the approach, rather than other things. Compared to the overall merit, the first or the millionth time it's done should not carry that much weight.

The proposed signature is more for human/semantics, not for compiler/syntax. The current signatures would be less confusing if an argument takes different semantics at each and every different position; but it only takes on the exception semantics if it is the last in position - that is a confusion that can be addressed by a strongly typed alternative.

**Reply by prdoyle** [reply]
Fair. I just mean, when I'm using a logging API, I expect the first parameter to be the message. Sticking something else in front will make it less readable.

However, having said that, in most cases it will look like `error(e, "Message goes here")` (because people often call exceptions `e`). Even I can admit that's no big deal.  😄

**Reply by q3769** [reply]
BTW having Throwable as the first arg is not new for other logging APIs, e.g. 

- https://tinylog.org/v2/logging/#exceptions

or to toot my own horn:

- https://github.com/elf4j/elf4j-api#the-logger-api

**Reply by prdoyle** [reply]
Heheh ok I yield.

**Comment by MpoMp** [comment]
This has been the source of much confusion and some mistakes in many companies I've worked in.
Personally I disagree with this hidden spec detail, it makes matters complicated and error-prone.

But even if it has to be there, it should be stated on the method-level Javadoc as well.
